scipy/optimize/_direct.py

```diff
@@ -0,0 +1,182 @@
+    ----------
+    func : objective function
+        called as `func(x, *args)`; does not need to be defined everywhere,
+        raise an Exception where function is not defined
+    bounds : array_like, optional
+            ``(min, max)`` pairs for each element in ``x``, defining
+            the bounds on that parameter.
+    nvar : int, optional
+        Dimensionality of x (only needed if `bounds` is not defined)
+    eps : float, optional
+        Ensures sufficient decrease in function value when a new potentially
+        optimal interval is chosen.
+    maxfun : int, optional
+        Approximate upper bound on objective function evaluations.
+        Maximal allowed value is 90000 see documentation of Fortran library.
+    maxiter : int, optional
+        Maximum number of iterations.
+        Maximal allowed value is 6000 see documentation of Fortran library.
+    locally_biased : bool, optional
+        Whether to use the original or modified DIRECT algorithm.
+        Possible values:
+
+        * ``locally_based=False`` - use the original DIRECT algorithm
+        * ``locally_biased=True`` - use the modified DIRECT-l algorithm
+    fglobal : float, optional
+        Function value of the global optimum.
+        By default it is a negative value whose absolute value is very large.
+        Set this value only if the global optimum is known.
+    fglper : float, optional
+        Terminate the optimization when the percent error satisfies:
+
+        .. math::
+
+            100*(f_{min} - f_{global})/\max(1, |f_{global}|) \leq f_{glper}
+    volper : float, optional
+        Terminate the optimization once the volume of a hyperrectangle is less
+        than volper percent of the original hyperrectangel.
+    sigma_per : float, optional
+        Terminate the optimization once the measure of the 
+        hyperrectangle is less than this argument.
+
+    Returns
+    -------
+    res : OptimizeResult
+        The optimization result represented as a ``OptimizeResult`` object.
+        Important attributes are: ``x`` the solution array, ``success`` a
+        Boolean flag indicating if the optimizer exited successfully and
+        ``message`` which describes the cause of the termination. See
+        `OptimizeResult` for a description of other attributes.
+    """
+
+    if bounds is None:
+        l = np.zeros(nvar, dtype=np.float64)
+        u = np.ones(nvar, dtype=np.float64)
+    else:
+        bounds = np.asarray(bounds)
+        l = bounds[:, 0]
+        u = bounds[:, 1]
+
+    def _objective_wrap(x, iidata, ddata, cdata, n, iisize, idsize, icsize):
+        r"""
+        Wrap the python objective to comply with the signature required by the
+        Fortran library.
+
+        Returns the function value and a flag indicating whether function 
+        is defined. If function is not defined, it returns ``np.nan``.
+        """
+        try:
+            return func(x, *args), 0
+        except (NameError, TypeError):
```

Comment:
Add tests for user returning a number, np.inf, -np.inf, np.nan and raising Exceptions.